### PR TITLE
Refactored chunking to use latest GELF protocol

### PR DESCRIPTION
**Pull request by livingtocode**
- updated JUnit dependency
- required gelf protocol is now 0.9.6+

The chunking protocol in Graylog2 changed after version 0.9.5, so gelf4j was not compatible with 0.9.6+ servers anymore which led to messages being dropped if the compressed message size exceeded 8K in size.

This update is bringing the communication protocol to the latest 1.1 Gelf protocol standard and is structuring the driver a bit better

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
+  mavenCentral()
 }
 
 dependencies {
-  groovy 'org.codehaus.groovy:groovy:1.7.8'
-  groovy 'log4j:log4j:1.2.16'
-  testCompile group: 'junit', name: 'junit', version: '4.8.2'
+  compile 'org.codehaus.groovy:groovy:1.7.8'
+  compile 'log4j:log4j:1.2.16'
+  testCompile 'junit:junit:4.12'
```
Comment (livingtocode) [inline build.gradle]: Brought this up to a newer gradle version and used a newer JUnit dependency

File: src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy [hunk 2]
```diff
@@ -76,83 +72,33 @@ extends AppenderSkeleton {
...
-        int len = throwableAsString.length
-        if(fullMessage.length()){ // newline after the 'original log message' to have nicer formatting
-          fullMessage.append Layout.LINE_SEP
-        }
-        for (int i = 0; i < len && i <  this.getMaxLoggedLines(); i++) {
-          fullMessage.append throwableAsString[i]
-          fullMessage.append Layout.LINE_SEP
-        }
-      }
-    }
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: All of this and the code removed below has been moved into the `GelfMessageBuilder` to make it a bit more structured and remove duplicated code

File: src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy [hunk 3]
```diff
@@ -76,83 +72,33 @@ extends AppenderSkeleton {
...
-      mdcFields.each {
-        def mdcValue = loggingEvent.getMDC(it)
-        if (mdcValue != null) {
-          gelfMessage['_' + it] = mdcValue as String
-        }
+    mdcFields.each {
+      def mdcValue = loggingEvent.getMDC(it as String)
+      if (mdcValue != null) {
+        messageBuilder.setAdditionalField(
+          'mdc_' + ((it as String) =~ /([A-Z])/).replaceAll('_$1').toLowerCase(),
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: MDC fields get converted from camel-case into underscore separated lowercase names to follow the GELF standard naming

File: src/main/groovy/com/pstehlik/groovy/gelf4j/appender/GelfMessageBuilder.groovy [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    setField('version', GELF_VERSION)
+
+    if (appender.includeLocationInformation && loggingEvent.locationInformationExists()) {
+      setAdditionalField('code_location_file', loggingEvent.locationInformation.fileName)
+      setAdditionalField('code_location_line', loggingEvent.locationInformation.lineNumber)
+      setAdditionalField('code_location_method_name', loggingEvent.locationInformation.methodName)
+    }
+
+    setAdditionalField('logger_name', loggingEvent.loggerName)
+    setAdditionalField('thread_name', loggingEvent.threadName)
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/GelfMessageBuilder.groovy]: Added additional information which may prove helpful in debugging based on log messages

File: src/main/groovy/com/pstehlik/groovy/gelf4j/net/BytesHelper.groovy [hunk 5]
```diff
@@ -1,42 +0,0 @@
-package com.pstehlik.groovy.gelf4j.net
-
-/**
- * based on net.sf.hibernate.util.BytesHelper
- *
- * @author Philip Stehlik
- * @since 0.8
- */
-public final class BytesHelper {
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/net/BytesHelper.groovy]: This wasn't used anywhere anymore

File: src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfTransport.groovy [hunk 6]
```diff
@@ -41,76 +26,29 @@ class GelfTransport {
...
-   * @param sequenceCount The total number of chunks
-   * @return The prepared byte array
-   */
-  private byte[] createChunkedMessagePart(String messageId, Integer sequenceNo, Integer sequenceCount) {
-    sequenceNo = sequenceNo -1
-    Collection ret = []
-    ret << 30.byteValue()
-    ret << 15.byteValue()
-    messageId.getBytes('ISO-8859-1').each {
-      ret << it
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfTransport.groovy]: All of the code above was cleaned up and moved into `GelfChunkBuilder` to restructure the package generation

File: src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfTransport.groovy [hunk 7]
```diff
@@ -118,92 +56,15 @@ class GelfTransport {
...
-    String formatted = Integer.toHexString(intval)
-    formatted = formatted.substring(0, formatted.size() >= lastDigits ? lastDigits : formatted.size())
-    return formatted.padLeft(lastDigits, '0')
-  }
-
-  protected static String format(short shortval, int lastDigits) {
-    String formatted = Integer.toHexString(shortval)
-    formatted = formatted.substring(0, formatted.size() >= lastDigits ? lastDigits : formatted.size())
-    return formatted.padLeft(lastDigits, '0')
-  }
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfTransport.groovy]: This wasn't necessary anymore as the unique ID for message chunks is now based on a random UUID

File: src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfChunkBuilder.groovy [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    this.maxChunkSize = maxChunkSize
+    this.totalChunkCount = totalChunkCount
+    header = ByteBuffer.allocate(HEADER_LENGTH)
+
+      .put(0x1e.byteValue())                            // The GELF magic numbers
+      .put(0x0f.byteValue())
+      .putLong(UUID.randomUUID().mostSignificantBits)   // a new random Message GUID
+      .put(0.byteValue())                               // the current chunk number
+      .put(totalChunkCount.byteValue())                 // the total chunk count
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/net/GelfChunkBuilder.groovy]: The chunk headers following the new GLEF format, including smaller fields for the ID chunk number and chunk count

File: src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy (hunk 3 above)
Comment (michaelnielson-taulia) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: I don't love the idea of running a regex on every field prior to sending them off to graylog, could we do this translation through a map at initialization time?

Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: That's a good point, as the MDC properties that will be send over are configured and won't change we can cache the translated names. Let me fix that.

File: src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy [hunk 9]
```diff
@@ -226,19 +174,23 @@ extends AppenderSkeleton {
 
   public void setMdcFields(List mdcFields) {
     this.mdcFields = mdcFields
+    this.mdcFields.each {
+      this.translatedMdcNames.put(it as String, 'mdc_' + ((it as String) =~ /([A-Z])/).replaceAll('_$1').toLowerCase())
```
Comment (livingtocode) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: @michaelnielson-taulia The MDC property names are now translated on initialization and cached in a map, so there is no regex tax on logging anymore

Comment (michaelnielson-taulia) [inline src/main/groovy/com/pstehlik/groovy/gelf4j/appender/Gelf4JAppender.groovy]: :metal: thanks!

**Pull request merged**